File: Leetcode/Javascript/285. Inorder Successor in BST.js
/*
    If < than root, must be in the left tree or root;
    if(>= root), must in the right tree;
*/
var inorderSuccessor = function(root, p) {
    if(!root) return null;
    if(p.val < root.val) {
        var tmp = inorderSuccessor(root.left, p);
        return tmp?tmp:root;
    }
    return inorderSuccessor(root.right, p);
};

File: Leetcode/Three/285. Inorder Successor in BST.java
public class Solution {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        if(root == null) return null;
        if(p.val < root.val) {
            TreeNode leftCandidate = inorderSuccessor(root.left, p);
            if(leftCandidate != null) return leftCandidate;
            else return root;
        }
        return inorderSuccessor(root.right, p);
    }
}
